xformers/triton/k_mem_efficient_attention.py

```diff
@@ -0,0 +1,169 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All rights reserved.
+#
+# This source code is licensed under the BSD license found in the
+# LICENSE file in the root directory of this source tree.
+
+import math
+from typing import Optional
+
+import torch
+import triton
+import triton.language as tl
+
+
+# fmt: off
+@triton.jit
+def k_me_attention_fw(
+    OUT, MAXES, WEIGHTS,    # out ptr
+    Q, K, V,                # in ptrs
+    M, N, L,                # dims
+    stride_out, stride_maxes, stride_weights,
+    **META,
+):
+    # fmt: on
+
+    # extract metaparameters
+    BLOCK_M = META["BLOCK_M"]
+    BLOCK_N, BLOCK_L = META["BLOCK_N"], META["BLOCK_L"]
+    scale = META["SCALE"]
+
+    # *within groups*, programs are ordered in a column-major order
+    # row-id /col-id of the program in the *launch grid*
+    pid_m = tl.program_id(axis=0)
+    pid_n = tl.program_id(axis=1)
+
+    # now compute the ranges that each program will go through
+    rm = pid_m * BLOCK_M + tl.arange(0, BLOCK_M)
+    rn = pid_n * BLOCK_N + tl.arange(0, BLOCK_N)
+
+    # initialize and iteratively update accumulator
+    acc = tl.zeros((BLOCK_M, BLOCK_N), dtype=tl.float32)
+
+    # block level matrix multiplication.
+    # We fetch a block memory block from both inputs, matmul and accumulate, then repeat
+    i = 0
+    rl = tl.arange(0, BLOCK_L)
+    for _ in range(L, 0, -BLOCK_L):
+        rl_i = rl + i * BLOCK_L
+        q_ptrs = Q + rm[:, None] * L + rl_i[None, :]    # (BLOCK_M, BLOCK_L)
+        k_ptrs = K + rn[None, :] * L + rl_i[:, None]    # (BLOCK_L, BLOCK_N)
+
+        q = tl.load(q_ptrs, mask=((rm[:, None] < M) & (rl_i[None, :] < L)), other=0.0)    # (BLOCK_M, BLOCK_L)
+        q *= scale  # q /= sqrt(dim)
+
+        k = tl.load(k_ptrs, mask=((rl_i[:, None] < L) & (rn[None, :] < N)), other=0.0)    # (BLOCK_L, BLOCK_N)
+
+        acc += tl.dot(q, k).to(tl.float32)              # (BLOCK_M, BLOCK_N)
+        i += 1
+
+    # pick the local max, safeguard the incoming exponential
+    # save so that an eventual mismatch can be fixed
+    max_acc = tl.max(acc, axis=1)                       # (BLOCK_M)
+    max_ptrs = MAXES + pid_n * stride_maxes + rm        # (BLOCK_M)
+    tl.store(max_ptrs, max_acc, mask=(rm < M))
+
+    # exponentiate the neutralized results
+    exp_acc = tl.exp(acc - max_acc[:, None])            # (BLOCK_M, BLOCK_N)
+
+    # Now pre-compute exp_acc against V.
+    # We proceed per chunk over L, and save as we go
+    i = 0
+    rl = tl.arange(0, BLOCK_L)
+    for _ in range(L, 0, -BLOCK_L):
+        rl_i = rl + i * BLOCK_L
+
+        v_ptrs = V + rn[:, None] * L + rl_i[None, :]    # (BLOCK_N, BLOCK_L)
+        v = tl.load(v_ptrs, mask=((rn[:, None] < N) & (rl_i[None, :] < L)), other=0.0)
+
+        qkv = tl.dot(exp_acc, v).to(tl.float32)         # (BLOCK_M, BLOCK_L)
```

Comment:
@ptillet not having .to(tl.float32) means that this crashes for instance, it's not really obvious to me why